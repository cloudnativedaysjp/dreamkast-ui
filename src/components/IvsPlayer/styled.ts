import styled from 'styled-components'

export const IvsPlayerContainer = styled.div`
  position: relative;
  width: 100%;
  height: 0;
  padding-top: 60%;
`

export const IvsPlayerVideo = styled.video`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100px;
  z-index: 0;
`

export const OverLayContainer = styled.div`
  background-color: rgba(0, 0, 0, 0.5);
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 94%;
  z-index: 10;
  color: white;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`

export const ButtonContainer = styled.div`
  display: flex;
  justify-content: space-around;
  width: 80%;
  max-width: 36em;
`

export const PlayerButton = styled.button`
  width: 6.2em;
  height: 1.8em;
  border: 2px solid lightgray;
  border-radius: 10px;
  background: rgba(0, 0, 0, 0);
  color: white;
  font-size: 1.6em;
  &:hover {
    background-color: rgba(189, 195, 199, 0.5);
  }
`

export const TextContainer = styled.div`
  width: 80%;
  padding-bottom: 28px;
  max-width: 36em;
`

export const NextTitle = styled.div`
  font-size: 1.2em;
  font-weight: bold;
`
