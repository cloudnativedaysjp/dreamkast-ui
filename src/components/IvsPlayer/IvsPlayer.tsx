import React, { useEffect, useRef } from 'react'
import videojs, { VideoJsPlayer } from 'video.js'
import 'video.js/dist/video-js.css'

import * as Styled from './styled'
import * as CommonStyled from '../../styles/styled'

type Props = {
  playBackUrl?: string
  autoplay: boolean
}

declare function registerIVSTech(
  vjs: typeof videojs,
  config?: { wasmWorker: string; wasmBinary: string },
): void

export const IvsPlayer: React.FC<Props> = ({ playBackUrl, autoplay }) => {
  const playerRef = useRef<VideoJsPlayer>()
  const videoElement = useRef<HTMLVideoElement>(null)

  useEffect(() => {
    const script = document.createElement('script')
    script.src =
      'https://player.live-video.net/1.4.1/amazon-ivs-videojs-tech.min.js'
    document.body.appendChild(script)

    script.addEventListener('load', () => {
      if (!playBackUrl || !videoElement.current) return
      registerIVSTech(videojs)
      const player = videojs(
        videoElement.current,
        {
          techOrder: ['AmazonIVS'],
          autoplay: autoplay,
        },
        () => {
          player.src(playBackUrl)
        },
      )
      playerRef.current = player
    })

    return () => {
      if (playerRef.current) playerRef.current.dispose()
      document.body.removeChild(script)
    }
  }, [playBackUrl])

  return (
    <CommonStyled.Container>
      <Styled.IvsPlayerContainer>
        <Styled.IvsPlayerVideo
          ref={videoElement}
          className="video-js vjs-16-9 vjs-big-play-centered"
          controls
          autoPlay
          playsInline
        />
      </Styled.IvsPlayerContainer>
    </CommonStyled.Container>
  )
}
