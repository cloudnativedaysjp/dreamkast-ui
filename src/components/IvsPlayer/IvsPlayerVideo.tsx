import React, { useEffect, useRef } from 'react'
import videojs, { VideoJsPlayer } from 'video.js'
import 'video.js/dist/video-js.css'

import * as Styled from './styled'

type Props = {
  playBackUrl?: string | null
  autoplay: boolean
}

declare function registerIVSTech(
  vjs: typeof videojs,
  config?: { wasmWorker: string; wasmBinary: string },
): void

export const IvsPlayerVideo: React.FC<Props> = ({ playBackUrl, autoplay }) => {
  const playerRef = useRef<VideoJsPlayer>()
  const videoElement = useRef<HTMLVideoElement>(null)

  useEffect(() => {
    const script = document.createElement('script')
    script.src =
      'https://player.live-video.net/1.4.1/amazon-ivs-videojs-tech.min.js'
    document.body.appendChild(script)

    script.addEventListener('load', () => {
      if (!videoElement.current) return
      registerIVSTech(videojs)
      const player = videojs(
        videoElement.current,
        {
          techOrder: ['AmazonIVS'],
          autoplay: autoplay,
        },
        () => {
          console.log('Player is ready to use!')
          if (playBackUrl) player.src(playBackUrl)
        },
      )
      playerRef.current = player
    })

    return () => {
      if (playerRef.current) playerRef.current.dispose()
      document.body.removeChild(script)
    }
  }, [])

  useEffect(() => {
    if (!playerRef.current) {
      return
    }
    if (!playBackUrl) {
      if (playerRef.current?.currentSource().src) {
        playerRef.current?.reset()
      }
      return
    }
    if (playBackUrl === playerRef.current?.currentSource().src) {
      return
    }
    playerRef.current.src(playBackUrl)
    console.log(playerRef.current.currentSource())
  }, [playBackUrl])

  return (
    <Styled.IvsPlayerVideo
      ref={videoElement}
      className="video-js vjs-16-9 vjs-big-play-centered"
      controls
      autoPlay
      playsInline
      muted={false}
    />
  )
}
