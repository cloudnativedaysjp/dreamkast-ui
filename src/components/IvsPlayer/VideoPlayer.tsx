import React, { useEffect, useRef } from 'react'
import videojs from 'video.js'
import 'video.js/dist/video-js.css'
import * as Styled from './styled'

interface VideoPlayerProps {
  options: videojs.PlayerOptions
}

export const VideoPlayer: React.FC<VideoPlayerProps> = ({ options }) => {
  const videoRef = useRef<HTMLVideoElement | null>(null)
  const playerRef = useRef<videojs.Player | null>(null)

  useEffect(() => {
    console.log('VideoPlayer.tsx: useEffect')
    // console.log('VideoPlayer.tsx options: ' + JSON.stringify(options))
    // 確実にvideoRefが存在することを確認し、適切な型として扱う
    const videoElement = videoRef.current as HTMLVideoElement
    if (!videoElement) return

    // video.jsのプレイヤーを初期化
    playerRef.current = videojs(videoElement, options)
    // console.log('sources: ' + JSON.stringify(options.sources))
    playerRef.current?.src(options.sources as videojs.Tech.SourceObject[])
    // console.log(
    //   'VideoPlayer.tsx playerRef.current: ' + JSON.stringify(playerRef.current),
    // )
    playerRef.current?.src(options.sources as videojs.Tech.SourceObject[])
    playerRef.current?.autoplay(true)

    // コンポーネントがアンマウントされた時にプレイヤーを破棄
    return () => {
      // if (playerRef.current) {
      //   playerRef.current.dispose()
      //   console.log('VideoPlayer.tsx playerRef.current.dispose()')
      // }
    }
  }, [options])

  return (
    <div data-vjs-player>
      <Styled.IvsPlayerVideo
        ref={videoRef}
        className="video-js vjs-16-9 vjs-big-play-centered"
      />
    </div>
  )
}
