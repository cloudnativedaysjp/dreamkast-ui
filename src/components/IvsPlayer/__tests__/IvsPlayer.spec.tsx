import React from 'react'
import { IvsPlayer } from '../IvsPlayer'
import { render } from '@testing-library/react'

describe('IvsPlayer', () => {
  it('should render video player when onAir', () => {
    const screen = render(
      <IvsPlayer
        videoCommand={{
          status: 'onAir',
          playBackUrl:
            'https://2a78e417f923.us-west-2.playback.live-video.net/api/video/v1/us-west-2.607167088920.channel.QFdTdOPixOx2.m3u8',
        }}
      />,
    )
    expect(screen.asFragment()).toMatchSnapshot()
  })

  it('should render video player when archived', () => {
    const screen = render(
      <IvsPlayer
        videoCommand={{
          status: 'archived',
          playBackUrl:
            'https://example.cloudfront.net/medialive/cndt2022/talks/0001/0001.m3u8',
        }}
      />,
    )
    expect(screen.asFragment()).toMatchSnapshot()
  })

  it('should render preparing page', () => {
    const screen = render(<IvsPlayer videoCommand={{ status: 'preparing' }} />)
    expect(screen.asFragment()).toMatchSnapshot()
  })

  it('should render archiving page', () => {
    const screen = render(<IvsPlayer videoCommand={{ status: 'archiving' }} />)
    expect(screen.asFragment()).toMatchSnapshot()
  })

  it('should render not selected page', () => {
    const screen = render(
      <IvsPlayer videoCommand={{ status: 'notSelected' }} />,
    )
    expect(screen.asFragment()).toMatchSnapshot()
  })
})
