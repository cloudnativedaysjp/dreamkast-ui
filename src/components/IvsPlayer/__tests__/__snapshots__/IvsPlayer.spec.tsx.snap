// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IvsPlayer 1`] = `
<div
  className="sc-dlfnbm fbDFAt"
>
  <div
    className="sc-bdfBwQ iOawCe"
  >
    <video
      autoPlay={true}
      className="sc-gsTCUz bBTKKj video-js vjs-16-9 vjs-big-play-centered"
      controls={true}
      playsInline={true}
    />
  </div>
</div>
`;
